common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -7881,6 +7940,192 @@ static jlong NativeCrypto_SSL_SESSION_get_time(JNIEnv* env, jclass, jlong ssl_se
+    return result;
+}
+
+/**
+ * Sets the timeout on the SSL session.
+ */
+static jlong NativeCrypto_SSL_SESSION_set_timeout(JNIEnv* env, jclass, jlong ssl_session_address, jlong millis) {
+    SSL_SESSION* ssl_session = to_SSL_SESSION(env, ssl_session_address, true);
+    JNI_TRACE("ssl_session=%p NativeCrypto_SSL_SESSION_set_timeout", ssl_session);
+    if (ssl_session == nullptr) {
+        return 0;
+    }
+
+    // Convert to seconds
+    long timeout = millis / 1000;
+    return SSL_set_timeout(ssl_session, timeout);
+}
+
+/**
+ * Gets the timeout for the SSL session.
+ */
+static jlong NativeCrypto_SSL_SESSION_get_timeout(JNIEnv* env, jclass, jlong ssl_session_address) {
+    SSL_SESSION* ssl_session = to_SSL_SESSION(env, ssl_session_address, true);
+    JNI_TRACE("ssl_session=%p NativeCrypto_SSL_SESSION_get_timeout", ssl_session);
+    if (ssl_session == nullptr) {
+        return 0;
+    }
+
+    jlong result = SSL_get_timeout(ssl_session);
+    result *= 1000; // OpenSSL uses seconds, Java uses milliseconds.
+    JNI_TRACE("ssl_session=%p NativeCrypto_SSL_SESSION_get_timeout => %lld", ssl_session, (long long) result);
+    return result;
+}
+
+/**
+ * Gets the ID for the SSL session, or null if no session is currently available.
+ */
+static jbyteArray NativeCrypto_SSL_session_id(JNIEnv* env, jclass, jlong ssl_address) {
+    SSL* ssl = to_SSL(env, ssl_address, true);
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_session_id", ssl);
+    if (ssl == nullptr) {
+        return nullptr;
+    }
+
+    SSL_SESSION* ssl_session = SSL_get_session(ssl);
+    JNI_TRACE("ssl_session=%p NativeCrypto_SSL_session_id", ssl_session);
+    if (ssl_session == nullptr) {
+        return nullptr;
+    }
+
+    jbyteArray result = env->NewByteArray(static_cast<jsize>(ssl_session->session_id_length));
+    if (result != nullptr) {
+        jbyte* src = reinterpret_cast<jbyte*>(ssl_session->session_id);
+        env->SetByteArrayRegion(result, 0, static_cast<jsize>(ssl_session->session_id_length), src);
+    }
+    JNI_TRACE("ssl_session=%p NativeCrypto_SSL_session_id => %p session_id_length=%d",
+             ssl_session, result, ssl_session->session_id_length);
+    return result;
+}
+
+/**
+ * Gets and returns in a string the cipher negotiated for the SSL session.
+ */
+static jstring NativeCrypto_SSL_cipher(JNIEnv* env, jclass, jlong ssl_address) {
+    SSL* ssl = to_SSL(env, ssl_address, true);
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_cipher", ssl);
+    if (ssl == nullptr) {
+        return nullptr;
+    }
+
+    SSL_SESSION* ssl_session = SSL_get_session(ssl);
+    JNI_TRACE("ssl_session=%p NativeCrypto_SSL_cipher", ssl_session);
+    if (ssl_session == nullptr) {
+        return nullptr;
+    }
+
+    const SSL_CIPHER* cipher = ssl_session->cipher;
+    const char* name = SSL_CIPHER_get_name(cipher);
+    JNI_TRACE("ssl_session=%p NativeCrypto_SSL_cipher => %s", ssl_session, name);
+    return env->NewStringUTF(name);
```

Comment:
The whole point was to avoid `SSL_get_session` and direct accesses of `SSL_SESSION` fields. I meant in the C++ code *and* the Java code. If you need to add a call to `SSL_get_session`, there is probably something wrong. If you need to reach into a field directly, there is also probably something wrong.

Use BoringSSL's public APIs. Don't reach into library internals. This should be `SSL_get_current_cipher`.